### PR TITLE
Allow for more elemwise torch functions using `broadcast_tensor` and `vmap`

### PR DESCRIPTION
**Pull request by Ch0ronomato**
## Description
In the event the operator `Elemwise` is broadcasting over doesn't have a direct torch function, we can leverage `vmap` and `broadcast_tensors` to replicate the ufunc machinery. 

## Related Issue

- [ ] Closes #
- [ ] Related to #1031 

## Checklist

- [X] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [X] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [X] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pytensor--1032.org.readthedocs.build/en/1032/

### CODE REVIEW

**Comment by Ch0ronomato** [comment]
I need to add a test, but I want to get feedback on #1031 before continuing.

**Comment by Ch0ronomato** [comment]
I'll fix the tests.

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 1]
```diff
@@ -11,9 +11,25 @@ def pytorch_funcify_Elemwise(op, node, **kwargs):
...
+            return base_fn(*inputs)
+    else:
+
+        def elemwise_fn(*inputs):
+            Elemwise._check_runtime_broadcast(node, inputs)
+            shaped_inputs = torch.broadcast_tensors(*inputs)
+            if shaped_inputs[0].dim() == 1:
+                ufunc = torch.vmap(base_fn)
+            else:
+                dims = (tuple(range(shaped_inputs[0].dim())),)
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:28]: I don't think this is correct. You need to apply vmap repeatedly to make it vectorize across multiple dimensions

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/elemwise.py:28]: You're correct, my bad.

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 2]
```diff
@@ -11,9 +11,23 @@ def pytorch_funcify_Elemwise(op, node, **kwargs):
...
+    else:
+
+        def elemwise_fn(*inputs):
+            Elemwise._check_runtime_broadcast(node, inputs)
+            shaped_inputs = torch.broadcast_tensors(*inputs)
+            ufunc = base_fn
+            for _ in range(shaped_inputs[0].dim()):
+                ufunc = torch.vmap(ufunc)
+            # @todo: This will fail for anything that calls
+            # `.item()`
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:29]: We usually use `# TODO:` not `# @todo:`

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 3]
```diff
@@ -11,9 +11,23 @@ def pytorch_funcify_Elemwise(op, node, **kwargs):
...
-        return base_fn(*inputs)
+    if hasattr(scalar_op, "nfunc_spec") and hasattr(torch, scalar_op.nfunc_spec[0]):
+        # torch can handle this scalar
+        # broadcast, we'll let it.
+        def elemwise_fn(*inputs):
+            Elemwise._check_runtime_broadcast(node, inputs)
+            return base_fn(*inputs)
+    else:
+
+        def elemwise_fn(*inputs):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:22]: We need a test for this branch. You can create a new test ScalarOp and dispatch something specifically in PyTorch (in the test suite) to test this?

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/elemwise.py:22]: Added

File: tests/link/pytorch/test_elemwise.py [hunk 4]
```diff
@@ -150,3 +153,33 @@ def test_cast():
         fgraph, [np.arange(6, dtype="float32").reshape(2, 3)]
     )
     assert res.dtype == torch.int32
+
+
+def test_vmap_elemwise():
+    from pytensor.link.pytorch.dispatch.basic import pytorch_funcify
+
+    class TestOp(ScalarOp):
```
Comment (Ch0ronomato) [inline tests/link/pytorch/test_elemwise.py:161]: I'm not sure if this should be defined at the module level or function level. If you want me to move it out lmk

Comment (ricardoV94) [reply tests/link/pytorch/test_elemwise.py:161]: doesn't matter. Usually I put it inside the test if it's not reused

File: pytensor/link/pytorch/linker.py [hunk 5]
```diff
@@ -13,7 +13,11 @@ def input_filter(self, inp: Any) -> Any:
         return pytorch_typify(inp)
 
     def output_filter(self, var: Variable, out: Any) -> Any:
-        return out.cpu()
+        import torch
+
+        if torch.is_tensor(out):
+            return out.cpu()
+        return out
```
Comment (ricardoV94) [inline pytensor/link/pytorch/linker.py:20]: This whole thing is suboptimal. I think we don't need it for JAX (which returns JAX arrays/ not numpy arrays), because `np.asarray` works with it, and I guess it doesn't work for torch tensors. This should only be needed for updated shared variables where we have to convert to a common type as they could be used in multiple functions with distinct backends.

Perhaps we should expand a bit on the TorchLinker to perform the updates itself, and only force conversion when that's the case. This is already supported by `Function`.

Comment (ricardoV94) [reply pytensor/link/pytorch/linker.py:20]: This is not a blocker to this PR as the problem was already here before

**Review by ricardoV94 — COMMENTED** [review]
Only some nits left, PR looks great!

File: pytensor/link/pytorch/linker.py (hunk 5 above)
Comment (ricardoV94) [reply pytensor/link/pytorch/linker.py:20]: Opened an issue: https://github.com/pymc-devs/pytensor/issues/1052

File: pytensor/link/pytorch/dispatch/elemwise.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/elemwise.py:29]: nope i agree with you.

**Comment by Ch0ronomato** [comment]
I loosen the failing test in #988 ; this should be good after that.

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 6]
```diff
@@ -11,9 +11,21 @@ def pytorch_funcify_Elemwise(op, node, **kwargs):
     scalar_op = op.scalar_op
     base_fn = pytorch_funcify(scalar_op, node=node, **kwargs)
 
-    def elemwise_fn(*inputs):
-        Elemwise._check_runtime_broadcast(node, inputs)
-        return base_fn(*inputs)
+    if hasattr(scalar_op, "nfunc_spec") and hasattr(torch, scalar_op.nfunc_spec[0]):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:14]: Should do the same trick for `scipy.x` you did in another PR?

File: pytensor/link/pytorch/linker.py [hunk 7]
```diff
@@ -13,7 +13,11 @@ def input_filter(self, inp: Any) -> Any:
         return pytorch_typify(inp)
 
     def output_filter(self, var: Variable, out: Any) -> Any:
-        return out.cpu()
+        import torch
+
+        if torch.is_tensor(out):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/linker.py:18]: Put these changes in a separate commit (and squash the Elemwise ones)

File: pytensor/link/pytorch/dispatch/elemwise.py (hunk 6 above)
Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/elemwise.py:14]: ...but then I have to write a new test for this PR 😅

File: pytensor/link/pytorch/linker.py (hunk 7 above)
Comment (Ch0ronomato) [reply pytensor/link/pytorch/linker.py:18]: okay sounds good

Comment (Ch0ronomato) [reply pytensor/link/pytorch/linker.py]: I got rid of this commit.

**Review by ricardoV94 — APPROVED** [review]
